eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/BbodySpec.hs

```diff
@@ -0,0 +1,139 @@
+import Cardano.Ledger.Babbage.Core
+import Cardano.Ledger.BaseTypes (BlocksMade (..), ProtVer (..), SlotNo (..))
+import Cardano.Ledger.Block
+import Cardano.Ledger.Coin (Coin (..))
+import Cardano.Ledger.Conway.Rules (ConwayBbodyPredFailure (..))
+import Cardano.Ledger.Plutus (SLanguage (..))
+import Cardano.Ledger.SafeHash (originalBytesSize)
+import Cardano.Ledger.Shelley.LedgerState
+import Cardano.Ledger.Shelley.Rules (
+  BbodyEnv (..),
+  LedgerEnv (..),
+  ShelleyBbodyState (..),
+ )
+import Cardano.Ledger.TxIn
+import Control.Monad (forM)
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map as Map
+import qualified Data.Sequence.Strict as SSeq
+import qualified Data.Set as Set
+import Lens.Micro ((&), (.~), (^.))
+import Lens.Micro.Mtl
+import Test.Cardano.Ledger.Babbage.ImpTest
+import Test.Cardano.Ledger.Imp.Common
+import Test.Cardano.Ledger.Plutus.Examples (alwaysFailsWithDatum)
+
+spec ::
+  forall era.
+  ( AlonzoEraImp era
+  , BabbageEraTxBody era
+  , EraSegWits era
+  , InjectRuleFailure "BBODY" ConwayBbodyPredFailure era
+  ) =>
+  SpecWith (ImpTestState era)
+spec = describe "BBODY" $ do
+  it "BodyRefScriptsSizeTooBig" $ do
+    Just (script :: Script era) <- pure largeScript
+    let scriptSize = originalBytesSize script
+    let maxRefScriptSizePerTx = 200 * 1024
+    let maxRefScriptSizePerBlock = 1024 * 1024
+
+    -- Determine a number of transactions and a number of times the reference script
+    -- needs to be included as an input in each transaction,
+    -- in order for the total to exceed the maximum allowed refScript size per block,
+    -- while the refScript size per individual transaction doesn't exceed maxRefScriptSizePerTx
+    txScriptCounts <-
+      genNumAdditionsExceeding
+        scriptSize
+        maxRefScriptSizePerTx
+        maxRefScriptSizePerBlock
+
+    let expectedTotalRefScriptSize = sum $ map (scriptSize *) txScriptCounts
+    txs <- do
+      -- Instead of using the rootTxIn, we are creating an input for each transaction
+      -- that we subsequently need to submit,
+      -- so that we can submit them independently of each other.
+      txIns <-
+        forM
+          txScriptCounts
+          ( \n -> do
+              txIn <- mkTxIn
+              pure (n, txIn)
+          )
+      forM
+        txIns
+        ( \(n, txIn) -> do
+            tx <- mkTxWithNScripts txIn script n
+            (fixupFees >=> updateAddrTxWits) tx
+        )
+
+    let txSeq = toTxSeq @era $ SSeq.fromList txs
+    nes <- use impNESL
+    LedgerEnv {ledgerPp, ledgerAccount} <- impLedgerEnv nes
+    let ls = nes ^. nesEsL . esLStateL
+    kh <- freshKeyHash
+    let bhView =
+          BHeaderView
+            { bhviewID = kh
+            , bhviewBSize = fromIntegral $ bBodySize (ProtVer (eraProtVerLow @era) 0) txSeq
+            , bhviewHSize = 0
+            , bhviewBHash = hashTxSeq txSeq
+            , bhviewSlot = SlotNo 0
+            }
```

Comment:
We do have access to the current slot number. Not that import, but could be useful for similar tests in the future
```suggestion
    slotNo <- gets impLastTick
    let bhView =
          BHeaderView
            { bhviewID = kh
            , bhviewBSize = fromIntegral $ bBodySize (ProtVer (eraProtVerLow @era) 0) txSeq
            , bhviewHSize = 0
            , bhviewBHash = hashTxSeq txSeq
            , bhviewSlot = slotNo
            }
```

Reply:
Definitely, thanks :+1:

Reply:
Ok to expose `impLastTick` ?

Reply:
We have a getter lens exposed for it: `impLastTickG`

Reply:
Indeed, thanks.